text/0027-corejam.md

```diff
@@ -0,0 +1,706 @@
+
+This ensures that Validators do a strictly limited amount of work before knowing whether they will be rewarded and are able to discontinue and attempt other candidates earlier than would otherwise be the case. There is the possibility of wasting Coretime by processing Work Packages which result in error, but well-written authorization procedures can mitigate this risk by making a prior validation of the Work Items.
+
+### Refine
+
+The `refine` function is implemented as an entry-point inside a code blob which is stored on-chain and whose hash is associated with the Work Class.
+
+```rust
+type ClassCodeHash = StorageMap<ClassId, CodeHash>;
+```
+
+```rust
+type WorkOutputLen = ConstU32<1024>;
+type WorkOutput = BoundedVec<u8, WorkOutputLen>;
+fn refine(payload: WorkPayload) -> WorkOutput;
+```
+
+Both `refine` and `is_authorized` are only ever executed in-core. Within this environment, we need to ensure that we can interrupt computation not long after some well-specified limit and deterministically determine when an invocation of the VM exhausts this limit. Since the exact point at which interruption of computation need not be deterministic, it is expected to be executed by a streaming JIT transpiler with a means of approximate and overshooting interruption coupled with deterministic metering.
+
+Several host functions (largely in line with the host functions available to Parachain Validation Function code) are supplied. Two additional ones include:
+
+```rust
+/// Determine the preimage of `hash` utilizing the Relay-chain Storage pallet.
+fn lookup(hash: [u8; 32]) -> Vec<u8>;
+/// Determine the state root of the block at given `height`.
+fn state_root(height: u32) -> Option<[u8; 32]>;
+```
+
+Other host functions will allow for the possibility of executing a WebAssembly payload (for example, a Parachain Validation Function) or instantiating and entering a subordinate RISCV VM (for example for Actor Progressions).
+
+When applying `refine` from the client code, we must allow for the possibility that the VM exits unexpectedly or does not end. Validators are always rewarded for computing properly authorized Work Packages, including those which include such broken Work Items. But they must be able to report their broken state into the Relay-chain in order to collect their reward. Thus we define a type `WorkResult`:
+
+```rust
+enum WorkError {
+    Timeout,
+    Panic,
+}
+struct WorkResult {
+    class: WorkClass,
+    item_hash: [u8; 32],
+    result: Result<WorkOutput, WorkError>,
+    weight: Weight,
+}
+fn apply_refine(item: WorkItem) -> WorkResult;
+```
+
+The amount of weight used in executing the `refine` function is noted in the `WorkResult` value, and this is used later in order to help apportion on-chain weight (for the Join-Accumulate process) to the Work Classes whose items appear in the Work Packages.
+
+```rust
+struct WorkReport {
```

Comment:
This must commit to the erasure-root of the work payload (and preferably its size, as well).

Reply:
This would be better than the `WorkPackageHash` as we have at present?